### ISSUE TITLE
The subsequent optimizer steps in VQ Trainer are wrong :)

### ISSUE THREAD

**Issue by IoSonoMarco** [Shubhamai/pytorch-vqgan#3]
Even though your zero the D gradients, when you do the backward, the gradients of G are affected by the gradients of D(G(z)). Since you perform two optimizer steps subsequently, G is updated by considering the additional gradient of D.

**Comment by UkiTenzai** [comment]
@IoSonoMarco @Shubhamai Agree. To ensure orderly and stable updates during adversarial training, the generator needs to perform at least one forward pass, while the discriminator typically requires three forward passes. For example, it can be structured like this:

1. **Forward pass the generator** to produce fake samples.

2. **Forward pass the discriminator** with:

   * real samples **with `.detach()`**, and
   * generated samples **with `.detach()`**.
     Then **backpropagate and update the discriminator**.

   > ⚠️ Note: `.detach()` is **necessary** here to prevent gradients from flowing into the generator when updating the discriminator.

3. **Forward the generated samples (without `.detach()`) to the discriminator**, and use the gradients from the discriminator output **together with other losses** (e.g. perceptual loss, L1, etc.) to **backpropagate and update the generator**.

In the current implementation, the author attempts to retain the computation graph by using retain_graph=True during backpropagation, instead of applying detach() and performing a second backward pass for the generator. However, this approach can undermine the orderly and stable update between the generator and discriminator. Specifically, on more complex datasets (e.g., human faces), this may cause the discriminator's loss to drop too quickly, leading to ineffective updates of the generator.

The corresponding training code can be written as follows:
```
def step(self, imgs: torch.Tensor) -> torch.Tensor:
    """Performs a single training step from the dataloader images batch

    For the VQGAN, it calculates the perceptual loss, reconstruction loss, and the codebook loss and does the backward pass.

    For the discriminator, it calculates lambda for the discriminator loss and does the backward pass.

    Args:
        imgs: input tensor of shape (batch_size, channel, H, W)

    Returns:
        decoded_imgs: output tensor of shape (batch_size, channel, H, W)
    """

    # Getting decoder output
    decoded_images, _, q_loss = self.vqgan(imgs)

    """
    =======================================================================================================================
    update discriminator
    """
    disc_real = self.discriminator(imgs.detach())
    disc_fake = self.discriminator(decoded_images.detach())

    d_loss_real = torch.mean(F.relu(1.0 - disc_real))
    d_loss_fake = torch.mean(F.relu(1.0 + disc_fake))

    disc_factor = self.vqgan.adopt_weight(self.disc_factor, self.global_step, threshold=self.disc_start)
    disc_loss = disc_factor * 0.5 * (d_loss_real + d_loss_fake)

    self.opt_disc.zero_grad()
    disc_loss.backward()
    self.opt_disc.step()

    """
    =======================================================================================================================
    update generator
    """
    perceptual_loss = self.perceptual_loss(imgs, decoded_images)
    rec_loss = torch.abs(imgs - decoded_images)
    perceptual_rec_loss = (
        self.perceptual_loss_factor * perceptual_loss
        + self.rec_loss_factor * rec_loss
    )
    perceptual_rec_loss = perceptual_rec_loss.mean()

    disc_fake = self.discriminator(decoded_images)
    adv_g_loss = torch.mean(F.relu(1.0 - disc_fake))

    # lambda_g = self.vqgan.calculate_lambda(perceptual_rec_loss, adv_g_loss)
    g_loss = perceptual_rec_loss + q_loss + disc_factor * adv_g_loss

    self.opt_g.zero_grad()
    g_loss.backward()
    self.opt_g.step()



    # ======================================================================================================================
    # Tracking metrics

    self.tb_writer.add_scalar("Perceptual & Reconstruction loss", perceptual_rec_loss, global_step=self.global_step)
    self.tb_writer.add_scalar("G Loss", g_loss, global_step=self.global_step)
    self.tb_writer.add_scalar("D Loss", disc_loss, global_step=self.global_step)


    return decoded_images, g_loss, disc_loss
```

Other suggestions for improving training stability and performance:

1. **Learning rate**: The current learning rate may be too small. Setting it to `1e-4` is generally a better choice and often leads to faster and more stable convergence.

2. **Optimizer choice**: In GAN training, the choice of optimizer is crucial:

   * If training is unstable, try using **RMSProp**, which is known to improve convergence in such cases.
   * If training is already stable, switching to **AdamW** can further improve performance.
     For example, in GANs trained on **speech data**, which are typically more stable, **AdamW** is widely adopted.

3. **GAN loss function**: The **hinge GAN loss** is very effective and widely used.
   Another good option is the **least squares GAN (LSGAN) loss**.
   In my experiments, the difference between these two loss functions is minimal in most cases.
